cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,316 @@
+	f := func(fw godo.Firewall) bool {
+		return fw.Name == fm.workerFirewallName
+	}
+	klog.Info("filtering firewall list for the firewall that has the expected firewall name")
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, errFailedToRetrieveFirewallList
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the public access firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	targetFirewall := fm.fwCache.getCachedFirewall()
+	isEqual := false
+	// A locally cached firewall with matching rules means there is nothing to update.
+	if targetFirewall != nil && cmp.Equal(targetFirewall.InboundRules, svcInboundRules) {
+		isEqual = true
+	}
+
+	if isEqual {
+		return nil
+	}
+
+	// A locally cached firewall exists, but the inbound rules don't match the expected
+	// service inbound rules. So we need to use the locally cached firewall ID to attempt
+	// to update the firewall APIs representation of the firewall with the new rules.
+	//
+	// Then we update the local cache with the firewall returned from the Update request.
+	if targetFirewall != nil {
+		fr := fm.createFirewallRequest(fm.workerFirewallName, fm.workerFirewallTags, svcInboundRules)
+		currentFirewall, resp, err := fm.client.Firewalls.Update(ctx, fm.fwCache.getFirewallID(), fr)
+		if err != nil {
+			if resp == nil || resp.StatusCode != http.StatusNotFound {
+				return errFailedToAddInboundRules
+			}
+			// Firewall does not exist, so we need to create a new firewall with the
+			// updated inbound rules.
+			currentFirewall, err = fm.createFirewall(ctx, svcInboundRules)
+			if err != nil {
+				return errFailedToCreateFirewallOnUpdate
+			}
+			klog.Infof("successfully created firewall")
+		}
+		fm.fwCache.updateCache(currentFirewall)
+		return nil
+	}
+
+	// Check if the target firewall ID exists but in the case that CCM first starts up and the
+	// firewall ID does not exist yet, we have to retrieve one from the API.
+	if targetFirewall == nil {
+		currentFirewall, err := fm.Get(ctx)
+		if err != nil {
+			return err
+		}
+		if currentFirewall == nil {
+			klog.Infof("an existing firewall not found, we need to create one")
+			currentFirewall, err = fm.createFirewall(ctx, svcInboundRules)
+			if err != nil {
+				return err
+			}
+			klog.Infof("successfully created firewall")
+		}
+		fm.fwCache.updateCache(currentFirewall)
+	}
+	return nil
+}
+
+func (fm *firewallManagerOp) createFirewall(ctx context.Context, svcInboundRules []godo.InboundRule) (*godo.Firewall, error) {
+	fr := fm.createFirewallRequest(fm.workerFirewallName, fm.workerFirewallTags, svcInboundRules)
+	currentFirewall, _, err := fm.client.Firewalls.Create(ctx, fr)
+	return currentFirewall, err
+}
+
+func (fm *firewallManagerOp) createFirewallRequest(fwName string, fwTags []string, rules []godo.InboundRule) *godo.FirewallRequest {
```

Comment:
In all existing call sites for `createFirewallRequest`, we always pass in values for `fwName` and `fwTags` that come from the `firewallManagerOp` struct. In that case, we can omit these two parameters from the signature and reference them directly from `fm` in this method's implementation.